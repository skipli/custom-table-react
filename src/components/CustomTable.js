import { createMuiTheme, responsiveFontSizes } from "@material-ui/core/styles";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import TableContainer from "@material-ui/core/TableContainer";
import React, { Component } from "react";
import { compare } from "../utils/compare";

export default class CustomTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      chronological: false,
    };
  }

  renderTableBody = () => {
    if (this.props.tableData.length === 0) return;
    let tableData = this.sortTableData();
    return tableData.map((data) => {
      return (
        <TableRow data-testid="table-row" key={data.id}>
          <TableCell>{new Date(data.timestamp).toDateString()}</TableCell>
          <TableCell>{data.id}</TableCell>
          <TableCell>{data.diff[0].oldValue}</TableCell>
          <TableCell>{data.diff[0].newValue}</TableCell>
        </TableRow>
      );
    });
  };

  sortDate = () => {
    if (this.state.chronological) this.setState({ chronological: false });
    else this.setState({ chronological: true });
  };

  sortTableData = () => {
    const sortedData = this.props.tableData.concat().sort((a, b) => {
      return compare(a.timestamp, b.timestamp);
    });
    if (!this.state.chronological) {
      return sortedData.reverse();
    } else return sortedData;
  };

  render() {
    let theme = createMuiTheme({
      overrides: {
        MuiTableCell: {
          head: {
            fontWeight: 600,
          },
        },
      },
    });
    responsiveFontSizes(theme);
    return (
      <TableContainer data-testid="table-container">
        <Table>
          <TableHead>
            <TableRow>
              <TableCell onClick={this.sortDate} align="left">
                Date
              </TableCell>
              <TableCell align="left">User ID</TableCell>
              <TableCell align="left">Old Value</TableCell>
              <TableCell align="left">New Value</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>{this.renderTableBody()}</TableBody>
        </Table>
      </TableContainer>
    );
  }
}
