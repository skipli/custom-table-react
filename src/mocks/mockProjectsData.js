export const mockProjectsDiff = {
  code: 200,
  data: [
    {
      id: "01fa742c-b2d0-416a-9394-a0a5dc8bc13e",
      timestamp: 1581717600000,
      diff: [
        {
          field: "name",
          oldValue: "Angrier Birds",
          newValue: "Angry Birds 2",
        },
      ],
    },
    {
      id: "7ba38998-0247-4de9-a3d8-e20cf736e834",
      timestamp: 1581631200000,
      diff: [
        {
          field: "name",
          oldValue: "Angry Birds",
          newValue: "Angrier Birds",
        },
      ],
    },
    {
      id: "c72d77f1-6e0b-473e-a600-7295d0822e54",
      timestamp: 1581804000000,
      diff: [
        {
          field: "name",
          oldValue: "Angry Birds 2",
          newValue: "Angry Birds 2: The sequel",
        },
      ],
    },
  ],
  limit: 3,
  offset: 0,
  total: 5,
};

export const mockProjectsDiffError = {
  code: 500,
  error: "Uknown error",
};
