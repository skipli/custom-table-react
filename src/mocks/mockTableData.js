export const mockTableData = [
  {
    id: "e28d290a-a2f2-48c2-9001-ff43884e271b",
    timestamp: 1581631200000,
    diff: [
      {
        field: "name",
        oldValue: "John",
        newValue: "Bruce",
      },
    ],
  },
  {
    id: "8bd0166f-f0c6-48fd-9fcd-a17e76eb1e92",
    timestamp: 1581717600000,
    diff: [
      {
        field: "name",
        oldValue: "Bruce",
        newValue: "Nick",
      },
    ],
  },
  {
    id: "0a75a2b3-be64-4aeb-ba4c-8ddb913791ac",
    timestamp: 1581804000000,
    diff: [
      {
        field: "name",
        oldValue: "Nick",
        newValue: "Michel",
      },
    ],
  },
];
