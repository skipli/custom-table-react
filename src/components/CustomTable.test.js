import React from "react";
import { shallow } from "enzyme";
import CustomTable from "./CustomTable";
import { mockTableData } from "../mocks/mockTableData";

describe("<CustomTable />", () => {
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = shallow(<CustomTable {...{ tableData: mockTableData }} />);
    instance = wrapper.instance();
  });

  describe("render()", () => {
    it("should be instance of the component", () => {
      expect(instance instanceof CustomTable).toBeTruthy();
    });

    it("should render the table", () => {
      expect(wrapper.find({ "data-testid": "table-container" })).toHaveLength(
        1
      );
    });
    it("should render table body rows from data", () => {
      expect(wrapper.find({ "data-testid": "table-row" })).toHaveLength(3);
    });
  });
});
