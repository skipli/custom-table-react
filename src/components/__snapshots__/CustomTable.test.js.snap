// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomTable /> render() should be instance of the component 1`] = `ShallowWrapper {}`;
